sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -545,29 +545,73 @@ ssize_t coap_tree_handler(coap_pkt_t *pkt, uint8_t *resp_buf, unsigned resp_buf_
     return coap_build_reply(pkt, COAP_CODE_404, resp_buf, resp_buf_len, 0);
 }
 
+ssize_t coap_build_reply_header(coap_pkt_t *pkt, unsigned code,
+                                void *buf, size_t len,
+                                int ct,
+                                void **payload, size_t *payload_len_max)
+{
+    uint8_t *bufpos = buf;
+    uint32_t no_response;
+    uint8_t type = coap_get_type(pkt) == COAP_TYPE_CON
+                 ? COAP_TYPE_ACK
+                 : COAP_TYPE_NON;
+
+    bufpos += coap_build_hdr(buf, type, coap_get_token(pkt), coap_get_token_len(pkt),
+                             code, ntohs(pkt->hdr->id));
+
+    if (coap_opt_get_uint(pkt, COAP_OPT_NO_RESPONSE, &no_response) == 0) {
+        const uint8_t no_response_index = (code >> 5) - 1;
+        /* If the handler code misbehaved here, we'd face UB otherwise */
+        assume(no_response_index < 7);
+
+        const uint8_t mask = 1 << no_response_index;
+        if (no_response & mask) {
+            if (payload) {
+                *payload = NULL;
+                *payload_len_max = 0;
+                payload = NULL;
+            }
+            if (type != COAP_TYPE_ACK) {
+                return 0;
+            }
+        }
+    }
+
+    if (payload) {
+        if (ct >= 0) {
+            bufpos += coap_put_option_ct(bufpos, 0, ct);
+        }
+        *bufpos++ = 0xff;
+        *payload = bufpos;
+        *payload_len_max = len - (bufpos - (uint8_t *)buf);
+    }
+
+    return bufpos - (uint8_t *)buf;
+}
+
 ssize_t coap_reply_simple(coap_pkt_t *pkt,
                           unsigned code,
                           uint8_t *buf, size_t len,
                           unsigned ct,
                           const void *payload, size_t payload_len)
 {
-    uint8_t *payload_start = buf + coap_get_total_hdr_len(pkt);
-    uint8_t *bufpos = payload_start;
+    void *payload_start;
+    size_t payload_len_max;
 
-    if (payload_len) {
-        bufpos += coap_put_option_ct(bufpos, 0, ct);
-        *bufpos++ = 0xff;
+    size_t header_len = coap_build_reply_header(pkt, code, buf, len, ct,
```

Comment:
```suggestion
    ssize_t header_len = coap_build_reply_header(pkt, code, buf, len, ct,
```

Also, you don't check for errors.